From 047f85b312b80499d6fd4dc4f23e3563eca92218 Mon Sep 17 00:00:00 2001
From: Janmejay Singh <singh.janmejay@gmail.com>
Date: Thu, 10 Mar 2016 00:20:49 +0530
Subject: [PATCH] fixes incorrect formatting(which leads to ill-formed json
 being generated) of stats when CEE/Json format is used

---
 runtime/statsobj.c                     | 11 +++--------
 tests/Makefile.am                      |  8 +++++++-
 tests/no-dynstats-json.sh              | 14 ++++++++++++++
 tests/no-dynstats.sh                   | 14 ++++++++++++++
 tests/testsuites/no-dynstats-json.conf |  9 +++++++++
 tests/testsuites/no-dynstats.conf      |  9 +++++++++
 6 files changed, 56 insertions(+), 9 deletions(-)
 create mode 100755 tests/no-dynstats-json.sh
 create mode 100755 tests/no-dynstats.sh
 create mode 100644 tests/testsuites/no-dynstats-json.conf
 create mode 100644 tests/testsuites/no-dynstats.conf

diff --git a/runtime/statsobj.c b/runtime/statsobj.c
index 848a130..dc941e5 100644
--- a/runtime/statsobj.c
+++ b/runtime/statsobj.c
@@ -285,9 +285,9 @@ getStatsLineCEE(statsobj_t *pThis, cstr_t **ppcstr, const statsFmtType_t fmt, co
 	rsCStrAppendStrWithLen(pcstr, UCHAR_CONSTANT("\""), 1);
 	rsCStrAppendStr(pcstr, pThis->name);
 	rsCStrAppendStrWithLen(pcstr, UCHAR_CONSTANT("\""), 1);
-	rsCStrAppendStrWithLen(pcstr, UCHAR_CONSTANT(","), 1);
 
 	if(pThis->origin != NULL) {
+		rsCStrAppendStrWithLen(pcstr, UCHAR_CONSTANT(","), 1);
 		rsCStrAppendStrWithLen(pcstr, UCHAR_CONSTANT("\""), 1);
 		rsCStrAppendStrWithLen(pcstr, UCHAR_CONSTANT("origin"), 6);
 		rsCStrAppendStrWithLen(pcstr, UCHAR_CONSTANT("\""), 1);
@@ -295,13 +295,12 @@ getStatsLineCEE(statsobj_t *pThis, cstr_t **ppcstr, const statsFmtType_t fmt, co
 		rsCStrAppendStrWithLen(pcstr, UCHAR_CONSTANT("\""), 1);
 		rsCStrAppendStr(pcstr, pThis->origin);
 		rsCStrAppendStrWithLen(pcstr, UCHAR_CONSTANT("\""), 1);
-		rsCStrAppendStrWithLen(pcstr, UCHAR_CONSTANT(","), 1);
 	}
 
 	/* now add all counters to this line */
 	pthread_mutex_lock(&pThis->mutCtr);
 	for(pCtr = pThis->ctrRoot ; pCtr != NULL ; pCtr = pCtr->next) {
-		rsCStrAppendStrWithLen(pcstr, UCHAR_CONSTANT("\""), 1);
+		rsCStrAppendStrWithLen(pcstr, UCHAR_CONSTANT(",\""), 2);
 		if (fmt == statsFmt_JSON_ES) {
 			/* work-around for broken Elasticsearch JSON implementation:
 			 * we need to replace dots by a different char, we use bang.
@@ -328,14 +327,10 @@ getStatsLineCEE(statsobj_t *pThis, cstr_t **ppcstr, const statsFmtType_t fmt, co
 			rsCStrAppendInt(pcstr, *(pCtr->val.pInt));
 			break;
 		}
-		if (pCtr->next != NULL) {
-			cstrAppendChar(pcstr, ',');
-		} else {
-			cstrAppendChar(pcstr, '}');
-		}
 		resetResettableCtr(pCtr, bResetCtrs);
 	}
 	pthread_mutex_unlock(&pThis->mutCtr);
+	cstrAppendChar(pcstr, '}');
 
 	CHKiRet(cstrFinalize(pcstr));
 	*ppcstr = pcstr;
diff --git a/tests/Makefile.am b/tests/Makefile.am
index 8e01271..e795ca2 100644
--- a/tests/Makefile.am
+++ b/tests/Makefile.am
@@ -243,7 +243,9 @@ TESTS +=  \
 	dynstats_overflow.sh \
 	dynstats_reset.sh \
 	dynstats_ctr_reset.sh \
-	dynstats_nometric.sh
+	dynstats_nometric.sh \
+	no-dynstats-json.sh \
+	no-dynstats.sh
 if HAVE_VALGRIND
 TESTS +=  \
 	dynstats-vg.sh \
@@ -985,6 +987,10 @@ EXTRA_DIST= \
 	testsuites/dynstats_nometric.conf \
 	testsuites/dynstats_overflow.conf \
 	testsuites/dynstats_reset.conf \
+	no-dynstats-json.sh \
+	testsuites/no-dynstats-json.conf \
+	no-dynstats.sh \
+	testsuites/no-dynstats.conf \
 	mmnormalize_variable.sh \
 	mmnormalize_tokenized.sh \
 	testsuites/mmnormalize_variable.conf \
diff --git a/tests/no-dynstats-json.sh b/tests/no-dynstats-json.sh
new file mode 100755
index 0000000..e2ef6ff
--- /dev/null
+++ b/tests/no-dynstats-json.sh
@@ -0,0 +1,14 @@
+#!/bin/bash
+# added 2016-03-10 by singh.janmejay
+# This file is part of the rsyslog project, released under ASL 2.0
+echo ===============================================================================
+echo \[no-dynstats-json.sh\]: test for verifying stats are reported correctly in json format in absence of any dynstats buckets being configured
+. $srcdir/diag.sh init
+. $srcdir/diag.sh startup no-dynstats-json.conf
+. $srcdir/diag.sh wait-for-stats-flush 'rsyslog.out.stats.log'
+echo doing shutdown
+. $srcdir/diag.sh shutdown-when-empty
+echo wait on shutdown
+. $srcdir/diag.sh wait-shutdown
+. $srcdir/diag.sh custom-content-check '{"name":"global","origin":"dynstats"}' 'rsyslog.out.stats.log'
+. $srcdir/diag.sh exit
diff --git a/tests/no-dynstats.sh b/tests/no-dynstats.sh
new file mode 100755
index 0000000..3144b70
--- /dev/null
+++ b/tests/no-dynstats.sh
@@ -0,0 +1,14 @@
+#!/bin/bash
+# added 2016-03-10 by singh.janmejay
+# This file is part of the rsyslog project, released under ASL 2.0
+echo ===============================================================================
+echo \[no-dynstats.sh\]: test for verifying stats are reported correctly in legacy format in absence of any dynstats buckets being configured
+. $srcdir/diag.sh init
+. $srcdir/diag.sh startup no-dynstats.conf
+. $srcdir/diag.sh wait-for-stats-flush 'rsyslog.out.stats.log'
+echo doing shutdown
+. $srcdir/diag.sh shutdown-when-empty
+echo wait on shutdown
+. $srcdir/diag.sh wait-shutdown
+. $srcdir/diag.sh custom-content-check 'global: origin=dynstats' 'rsyslog.out.stats.log'
+. $srcdir/diag.sh exit
diff --git a/tests/testsuites/no-dynstats-json.conf b/tests/testsuites/no-dynstats-json.conf
new file mode 100644
index 0000000..9a09333
--- /dev/null
+++ b/tests/testsuites/no-dynstats-json.conf
@@ -0,0 +1,9 @@
+$IncludeConfig diag-common.conf
+
+ruleset(name="stats") {
+  action(type="omfile" file="./rsyslog.out.stats.log")
+}
+
+module(load="../plugins/impstats/.libs/impstats" interval="1" severity="7" resetCounters="on" Ruleset="stats" bracketing="on" format="json")
+
+action(type="omfile" file="./rsyslog.out.log")
\ No newline at end of file
diff --git a/tests/testsuites/no-dynstats.conf b/tests/testsuites/no-dynstats.conf
new file mode 100644
index 0000000..7ee04b9
--- /dev/null
+++ b/tests/testsuites/no-dynstats.conf
@@ -0,0 +1,9 @@
+$IncludeConfig diag-common.conf
+
+ruleset(name="stats") {
+  action(type="omfile" file="./rsyslog.out.stats.log")
+}
+
+module(load="../plugins/impstats/.libs/impstats" interval="1" severity="7" resetCounters="on" Ruleset="stats" bracketing="on")
+
+action(type="omfile" file="./rsyslog.out.log")
\ No newline at end of file
